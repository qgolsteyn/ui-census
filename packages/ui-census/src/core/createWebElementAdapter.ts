import createAsyncAdapter from "./createAsyncAdapter";

import { CensusDefinitionAsync, CensusObjectAsync } from "../types";
import { WebElement } from "selenium-webdriver";

// We re-export the adapter factory with the correct type
// for use with DOM or JSDOM elements.

const createWebElementAdapter: <Definition extends CensusDefinitionAsync<
  WebElement
>>(
  definition: Definition
) => (target: WebElement) => CensusObjectAsync<Definition> = createAsyncAdapter;

export default createWebElementAdapter;
