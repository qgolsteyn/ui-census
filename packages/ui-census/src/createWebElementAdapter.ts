import createAsyncAdapter from "./createAsyncAdapter";

import {
  CensusObject,
  CensusDefinitionAsync,
  CensusObjectAsync,
} from "./types";
import { WebElement, By } from "selenium-webdriver";

// We re-export the adapter factory with the correct type
// for use with DOM or JSDOM elements.

const createWebElementAdapter: <Definition extends CensusDefinitionAsync<
  WebElement
>>(
  definition: Definition
) => (target: WebElement) => CensusObjectAsync<Definition> = createAsyncAdapter;

export default createWebElementAdapter;
