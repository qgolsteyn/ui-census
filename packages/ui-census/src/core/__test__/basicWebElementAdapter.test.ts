import { Builder, By } from "selenium-webdriver";
import chrome from "selenium-webdriver/chrome";

import createWebElementAdapter from "../createWebElementAdapter";

import "chromedriver";

const b = createWebElementAdapter({
  h1: {
    selector: (target) => target.findElements(By.css("h1")),
    queries: {
      text: (element) => element.getText(),
    },
    actions: {},
  },
  p: {
    selector: (target) => target.findElements(By.css("p")),
    queries: {
      text: (element) => element.getText(),
    },
    actions: {},
  },
});

const options = new chrome.Options();
options.addArguments("headless");

const driver = new Builder()
  .forBrowser("chrome")
  .setChromeOptions(options)
  .build();
driver.get("http://example.com");

const doc = b(driver.findElement(By.css("body")));

afterAll(async () => {
  await driver.quit();
});

test("it handles a basic query", async () => {
  const h1 = await doc.h1().single();
  expect(h1.text).toBe("Example Domain");
});

test("it handles a complex query", async () => {
  const p = await doc.p().all();
  expect(p.map((item) => item.text)).toEqual([
    "This domain is for use in illustrative examples in documents. You may use this domain in literature without prior coordination or asking for permission.",
    "More information...",
  ]);
});

test("it handles a simple query", async () => {
  const match = await doc.h1().contains({ text: "Example Domain" }).single();

  expect(match).toEqual({ text: "Example Domain" });
});
test("it handles snapshot testing", async () => {
  const p = await doc.p().all();
  expect(p).toMatchSnapshot();
});
