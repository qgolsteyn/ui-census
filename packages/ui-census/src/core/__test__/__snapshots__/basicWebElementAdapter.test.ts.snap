// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it handles snapshot testing 1`] = `
Array [
  Object {
    "text": "This domain is for use in illustrative examples in documents. You may use this domain in literature without prior coordination or asking for permission.",
  },
  Object {
    "text": "More information...",
  },
]
`;
